src/dataflow/src/render/reduce.rs

```diff
@@ -208,144 +248,267 @@ where
+                if monotonic {
+                    LIrReduceExpr::Hierarchical(HierarchicalReduceExpr::Monotonic(aggregates_list))
+                } else {
+                    let mut shifts = vec![];
+                    let mut current = 4usize;
+
+                    // Plan for 4B records in the expected case
+                    let limit = expected_group_size.unwrap_or(4_000_000_000);
+
+                    while (1 << current) < limit {
+                        shifts.push(current);
+                        current += 4;
+                    }
+
+                    shifts.reverse();
+
+                    let regular = RegularHierarchicalReduceExpr {
+                        aggrs: aggregates_list,
+                        shifts,
+                    };
+
+                    LIrReduceExpr::Hierarchical(HierarchicalReduceExpr::Regular(regular))
+                }
+            }
+            ReductionType::Basic => {
+                if aggregates_list.len() == 1 {
+                    LIrReduceExpr::Basic(BasicReduceExpr::Single(
+                        aggregates_list[0].0,
+                        aggregates_list[0].1.clone(),
+                    ))
+                } else {
+                    LIrReduceExpr::Basic(BasicReduceExpr::Multiple(aggregates_list))
+                }
+            }
+        }
+    };
+
+    if aggregates.is_empty() {
+        return LIrReduceExpr::Distinct;
+    }
+
+    let mut reduction_types = BTreeMap::new();
+    // We need to make sure that each list of aggregates by type forms
+    // a subsequence of the overall sequence of aggregates.
+    for index in 0..aggregates.len() {
+        let typ = reduction_type(&aggregates[index].func);
+        let aggregates_list = reduction_types.entry(typ).or_insert(Vec::new());
+        aggregates_list.push((index, aggregates[index].clone()));
+    }
+
+    let lir: Vec<_> = reduction_types
+        .into_iter()
+        .map(|(typ, aggregates_list)| lower(typ, aggregates_list))
+        .collect();
+
+    if lir.len() == 1 {
+        return lir[0].clone();
+    }
+
+    assert!(lir.len() <= 3);
+    let mut collation: CollationReduceExpr = Default::default();
+    let aggregate_types = aggregates
+        .iter()
+        .map(|a| reduction_type(&a.func))
+        .collect::<Vec<_>>();
+
+    collation.aggregate_types = aggregate_types;
+
+    for expr in lir.into_iter() {
+        match expr {
+            LIrReduceExpr::Accumulable(e) => {
+                assert!(collation.accumulable.is_none());
+                collation.accumulable = Some(e);
+            }
+            LIrReduceExpr::Hierarchical(e) => {
+                assert!(collation.hierarchical.is_none());
+                collation.hierarchical = Some(e);
+            }
+            LIrReduceExpr::Basic(e) => {
+                assert!(collation.basic.is_none());
+                collation.basic = Some(e);
+            }
+            _ => panic!("Inner reduce LIR expr was unsupported type!"),
```

Comment:
this panic follows logically from the fact that `lower` only generates one of those three variants but the error message could definitely be better / maybe theres a better way to express the logic